components/sup/src/manager/service/mod.rs

```diff
@@ -935,14 +889,17 @@ impl Service {
             }
         } else {
             self.check_process();
-            // NOTE: if you need reconfiguration and you DON'T have a
-            // reload script, you're going to restart anyway.
-            if self.needs_reload || self.process_down() || self.needs_reconfiguration {
-                self.reload(launcher);
-                if self.needs_reconfiguration {
-                    // NOTE this only runs the hook if it's defined
-                    self.reconfigure()
-                }
+            if hook_update_table.run
+               || hook_update_table.post_run
+               || self.process_down()
+               || (self.hooks.reconfigure.is_none() && config_change)
+            {
+                // TODO (DM): This flag is a hack. We have the `TaskExecutor` here. We could just
+                // schedule the `stop` future, but the `Manager` wraps the `stop` future with
+                // additional functionality. Can we refactor to make this flag unnecessary?
+                self.needs_restart = true;
+            } else if config_change || hook_update_table.reconfigure {
```

Comment:
I _think_ this is correct, but the way the overall if/else conditional is structured makes it a little tricky to confirm without whipping up a truth table. Restructuring it so it's more obviously correct and complete would be good (that may involve some helper methods on the `hook_update_table` struct [mentioned earlier], some additional structure to let the compiler lend a hand here, etc.)